### PR TITLE
Docker backend

### PR DESCRIPTION
**Pull request by MisterDA**
This PR implements the Docker backend for OBuilder. It should support running on Windows with Docker for Windows, and on Linux with the usual Docker.
This PR contains a batch of prerequisites, and then the backend.
I have tested it on Windows and Linux, on basic examples.

- This PR requires some patches to Lwt. Some have been merged to the development branch already.
  + https://github.com/ocsigen/lwt/pull/869
  + https://github.com/ocsigen/lwt/pull/886
  + https://github.com/ocsigen/lwt/pull/898
  + https://github.com/ocsigen/lwt/pull/901
- [ ] Test the OBuilder Docker backend with ocaml-ci, opam-health-check, and opam-repo-ci.

### CODE REVIEW

**Review by talex5 — COMMENTED** [review]
There shouldn't be any docker-specific stuff in build.ml. Can the sandbox API be extended to allow making all this generic?

**Comment by talex5** [comment]
Thinking on about this... the Build module creates a builder from a store and a sandbox. That will work on Linux (runc+btrfs), mac (users+zfs) and probably on FreeBSD (jails+zfs). But it doesn't work with Docker, because Docker already combines a store and a sandbox. So probably this should be an alternative to the build.ml module instead.

**Comment by MisterDA** [comment]
Thanks for the review, I agree with your conclusion and will refactor the code to merge the Docker sandbox, `Docker_storage`, the Docker-related code in `build.ml`, and maybe the code to talk to the client in the `Docker` module into a single `Docker_build` module.

A note regarding the use of the [Docker Engine API](https://docs.docker.com/engine/api/): I wanted to replace the process calls to the `docker` client with the Docker Engine API, talking directly to `dockerd` with sockets or pipes. However
- [ocaml-docker](https://github.com/Chris00/ocaml-docker) supports API 1.29 and would need to be ported to Windows too;
- [openapi-generator](https://github.com/OpenAPITools/openapi-generator), a tool that generates bindings from a standardized yaml description of the API doesn't support the latest version of the standard and generates invalid OCaml code (see https://github.com/OpenAPITools/openapi-generator/issues/9787).

I'm putting this on hold for now, and I'll stay with the docker client until openapi-generator is fixed.

**Comment by MisterDA** [comment]
Now depends on 
- #116 
- #117

File: lib/docker_store.ml [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+  root : string;        (* The top-level directory (containing `result`, etc). *)
+  caches : (string, cache) Hashtbl.t;
+  mutable next : int;   (* Used to generate unique temporary IDs. *)
+}
+
+let ( / ) = Filename.concat
+let strf = Printf.sprintf
+
+module Path = struct
+  (* A btrfs store contains several subdirectories:
```
Comment (tmcgilchrist) [inline lib/docker_store.ml:24]: This should be a `Docker volume` store rather than btrfs?

**Comment by MisterDA** [comment]
I lost the race to merge before macOS. The rebase wasn't easy, I've probably broken one or two new things in my code.

**Pull request closed without merging**